### PR TITLE
Verbose perplexity evaluation

### PR DESCRIPTION
**Pull request by OyvindTafjord**
@IanMagnusson Here's an attempt at some initial code for doing more detailed perplexity evaluation (sorry about the spurious MMLU thingy in there, but I'll just leave it in). 

This works by logging the full list of tokens and logits for each instance. Then at the end of the evaluation, there's a post-processing method in the task (called `process_extra_output`) which will mutate the output (delete the stored tokens/logits while adding the accumulated stats).

This is currently just hooked into the `run_lm_eval` function, so for the tango-integration it presumably needs similar updates there. Here's an example:

```
python catwalk/run_lm_eval.py  --model lm::pretrained=EleutherAI/pythia-160m,revision=step140000  \
--task_file eval-suite-2b.jsonl --full_output_file predictions.jsonl --metrics_file metrics.json \
--limit 2 --batch_size 4 --model_max_length 2048
```

where the `task_file` looks like:
```
{"name": "ppl_custom", "split": "validation", "task_options": {"detailed_output": true, "files": ["/Users/oyvindt/data/olmo/olmo-ppl-val-v2/c4_en/val"]}, 
"task_rename": "ppl_c4", "keep_instance_fields":["orig_file_name","source","subdomain"]}
```
(note the new `"detailed_output"` flag). This will produce a similar predictions.jsonl file as before, but with a new entry with sorted list (by average logit) of all the tokens seen:
```
"extra_output": 
   {"token_count_avg_logits_by_domain": 
      {"c4_en__": {"Gy": [1, -17.37446403503418], "ind": [1, -16.550533294677734], " Height": [1, -16.409221649169922], " dich": [1, -15.504566192626953], "0038": [1, -15.439532279968262], "Statistical": [2, -15.352715253829956],
```
(the entries are token count and average logit). 

The downside of this method is using quite a bit of (inefficient) memory for storing the tokens and logits during evaluation, but I don't think that will be an issue for our datasets (if so, we can turn these into more efficient arrays). 

This should be flexible in allowing other postprocessing steps, such as capturing particular spans etc, since you'll have access to all the tokens and logits at that point (although not after the evaluation is done, unless you want to save the full tokens and logits).

### CODE REVIEW

**Review by IanMagnusson — APPROVED** [review]
This looks great to me! Thanks so much Oyvind! I'll work on getting this adopted by Akshita's tango pipeline as well and then throw a bunch of runs at it to see if the memory demands are manageable.
